Added CLIs and WDL for python gCNV pipeline.

samuelklee opened the pull request.

mbabadi commented:
@samuelklee awh hehe... will get rid of that column

mbabadi commented:
@samuelklee gCNV integration tests and WDL tests pass locally. There are a number of issues:
- Somatic WDL tests fail on travis, and haven't tried them locally. Those tests were failing before I made changes -- it's unlikely to be related to my changes. Could you please take a look?
- Germline WDL tests are extremely slow on travis, perhaps there is no g++ and it is using python fallback? I don't know how to see the cromwell log files on travis. @lbergelson any pro tips?
- Python-related integration tests all fail on travis (though not locally w/ a properly setup python environment).
- I noticed that the python env in `Dockerfile` is based on `Miniconda2`. We strictly require Python 3 for gCNV.

samuelklee commented:
I’ll take a look at the somatic tests.  They should be OK, probably just something related to kebab case changes.

I think the WDL tests should be using the Docker, which has g++.  Travis machines might be slower?

Integration tests will need to be in the python test group.  Take a look at the python tests.

samuelklee commented:
OK, I think I figured out what was going on.  In 9b194a6 I changed PreprocessIntervals to drop intervals with all Ns.  But I must not have rebuilt the somatic PoNs and updated the copies in the large test resources.  I could've sworn that I tested the somatic pipeline locally, but perhaps I forgot to update the jar at some point.

Ideally, we should figure out some way to use the PoNs built by the panel WDL tests in the subsequent tests for the case/pair workflows.

samuelklee commented:
@mbabadi  I pushed new PoNs, I think that should get somatic tests to pass.  Sorry about that, hope you didn't go too far down the rabbit hole trying to debug!

If you need to make changes to the Docker or Travis environments, perhaps coordinate with @cmnbroad to make them directly in #3912, which is still open.

mbabadi commented:
@samuelklee haha I just figured it out and was about to write to you. Hopefully the tests will pass now. The changes I made to travis yml were experimental -- I am reverting them.

mbabadi commented:
(except for Miniconda2 -> Miniconda3)

mbabadi commented:
@samuelklee darn; the wgs pon related tests are still failing and the germline WDL exceeds the time limit (it completes in ~ 10 mins on gsa5, which is already quite slow, and ~ 3 mins on my laptop!). Something strange is going on. I wish there was an easy way to inspect the cromwell logs; perhaps there's a way to set cromwell logging to stdout.

mbabadi commented:
@samuelklee OK it is not a cromwell issue -- after your commit, the somatic denoising integration tests are failing locally too:
```java.lang.IllegalArgumentException: Sample intervals must be identical to the original intervals used to build the panel of normals.```

I wouldn't worry about it now. It is most likely related to test resource files.

samuelklee commented:
@mbabadi I think the somatic tests are fixed: https://travis-ci.org/broadinstitute/gatk/builds/313333659?utm_source=email&utm_medium=notification  Did you pull my commit?

Feel free to cut out some of the test cases if the WDL tests are too slow.  Especially now that there is no real distinction between WGS/WES besides -L (except for in PreprocessIntervals, which integration tests should cover).

samuelklee commented on scripts/cnv_wdl/germline/README.md:
```diff
@@ -1,86 +1,53 @@
\ No newline at end of file
+- ``CNVGermlineCohortWorkflow.cohort_entity_id`` -- Name of the cohort.  Will be used as a prefix for output filenames.
+- ``CNVGermlineCohortWorkflow.contig_ploidy_priors`` -- TSV file containing prior probabilities for the ploidy of each contig, with column headers: CONTIG_NAME, PLOIDY_PRIOR_0, PLOIDY_PRIOR_1, ...
+- ``CNVGermlineCohortWorkflow.gatk_docker`` -- GATK Docker image (e.g., ``broadinstitute/gatk:x.beta.x``).
```
Update this.

samuelklee commented:
Rebased now that sl_wgs_acnv, sl_wgs_acnv_headers, and sl_delete_legacy_cnv code are all in.  Phew!

samuelklee commented:
Rebased and squashed on top of sl_wgs_acnv_headers_docs.  Here is the log of squashed commits, for reference:

````
commit 3eda4b18888f38249be39f99901d8453a4de50d6
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Thu Dec 28 14:56:27 2017 -0500

    updated command lines for WDL tests for C29

commit 7ce1369943cce4ae9cfb5e96455d18d3960e9b77
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Thu Dec 28 13:30:21 2017 -0500

    use C29 and decrease gcnv_max_training_epochs

commit 68772cba486b44ebc8cf8bfc2b600c1e8a406c61
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Sun Dec 17 19:20:05 2017 +0330

    documentation update of GermlineCNVCaller and DetermineGermlineContigPloidy

commit c032281f8c43a80e4ec8cb96eb66397ad2acf9b7
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Fri Dec 15 18:14:35 2017 -0500

    Fixed imr kebab case in WDL, moved argument classes, removed GenomeLocParser, fixed up gCNV WDL readme.

commit be84a804f6ab6fbb815995db9c116d1db950ab8b
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 12 13:55:08 2017 -0500

    removed extra comma in gCNV Case WDL test JSON

commit cb379b866d425f12f5525ecb28ad0b636a528d44
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 12 12:58:59 2017 -0500

    added missing cpu parameters to gCNV Case WDL tasks

commit eed85f6c70f4a7f15e0765b5f15a1bf8541c151e
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 12 11:26:31 2017 -0500

    disabled some gCNV WDL tests

commit 6d8ca07fef41518b5b157fb9a214d4536c617156
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 12 10:54:54 2017 -0500

    fixed DenoiseReadCountsIntegrationTest files

commit adfbef12f2ab90f93b49a4f786979549648e1f22
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Mon Dec 11 02:22:56 2017 -0500

    removed CNV evaluation code from this branch

commit 18c8d31f39a1964474c5d7b12ee8cbfafc4ac9e2
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Sun Dec 10 00:19:58 2017 -0500

    GS VCF parser outputs dict for samples instead of list

commit b138be39cd8428342668ee6678079021006f983b
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Sun Dec 10 00:15:19 2017 -0500

    renaming

commit eab5c90b74b4eb6bd11acb0fd1e0fa58a3b5b0c7
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Fri Dec 8 16:23:40 2017 -0500

    exposed a global preemptible_attempts to gCNV workflows
    set OMP_NUM_THREADS and MKL_NUM_THREADS to the number of requested CPUs

commit ad6fe348d6a7896c169b2b0499e2a4bca34021ad
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Fri Dec 8 10:21:25 2017 -0500

    reverted log level in germline CNV tests

commit d9eb4e504baab834a9efc07cc3479176db2946ce
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Fri Dec 8 10:20:30 2017 -0500

    the proper python environment yml for mkl and open -- leads to orders of magnitude speedup!

commit fea6bf874e0b62262a3b1d239ce4d76792d5c416
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Fri Dec 8 09:31:43 2017 -0500

    revert

commit 456c53f88d01b603f4175d8896a0dac036af03f8
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Fri Dec 8 08:17:22 2017 -0500

    enabled openmp g++ linking in theano

commit e2afef14ddb957f2dbdea76fd783d3bfb8d7a64e
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Fri Dec 8 08:04:19 2017 -0500

    mkl

commit 43e2a65201286161fcd5bfe7dbb21ae888e19dac
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Fri Dec 8 06:56:20 2017 -0500

    added cpu argument for germline tasks

commit 4433a62c2173c7f29d0f264c084bbaf2f6738782
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Fri Dec 8 02:45:38 2017 -0500

    revert travis yml forks
    verbose logging germline wdl

commit ae05801e33c37b3bf2685fba202032a270804873
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Fri Dec 8 00:55:14 2017 -0500

    updated somatic PoNs for PreprocessIntervals drop Ns

commit cff64984d9fb42364001bda4c73d54cf68d85a5c
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Fri Dec 8 00:37:24 2017 -0500

    sudo travis yml

commit 89025941febd2089d426cfa1e0f0aa6a6712e2a9
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Fri Dec 8 00:23:22 2017 -0500

    travis/Docker config update (g++-6, Miniconda3)
    python test group assignment

commit 31f96398106c2b8577b8c25d110abea3ebe7f836
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Thu Dec 7 20:44:53 2017 -0500

    WDL test bugfix

commit 9b2fb820536ec355bea0256471bd093d547f5c99
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Thu Dec 7 20:20:36 2017 -0500

    update WDL test JSON files

commit e3d97644d1a2c40a5c364a96f8b67246154179c9
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Thu Dec 7 20:18:14 2017 -0500

    assertions in inference task base
    removed a ASCII > 128 character in log messages

commit 526cf92e623a3bbd5f9d375132b6ca046fc47620
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Thu Dec 7 20:03:04 2017 -0500

    redirect tqdm progress bar to python logger

commit 2e45bd30968b921fae225de3901fb97ece690b0c
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Thu Dec 7 19:45:49 2017 -0500

    more arg related fixes

commit bb89a3bb338d88199881e8aca65f656f2acd7c0a
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Thu Dec 7 19:41:20 2017 -0500

    arg related bugfixes in WDL, python, and java CLIs

commit 23569787ee2c8cc6c9227a44170cbbd02fe4427f
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Thu Dec 7 17:21:05 2017 -0500

    fixed issue with python boolean argparse (they use weird semantics)

commit ae841c9ed4cd9b2ca1ac0e9082d175ff8ea98298
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Thu Dec 7 16:44:02 2017 -0500

    shorter gCNV WDL tests

commit 5466b806e36df16cad2d045be074e7f9afec0957
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Thu Dec 7 16:38:15 2017 -0500

    fixed arg issues in somatic WDL
    exposed all missing args to java side
    major update to germline WDLs
    all optional python args exposed to WDLs as optional args

commit 50cb6fd08de15469a9080cbb27ff30c8b7ee7e21
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Thu Dec 7 13:50:45 2017 -0500

    missing serialVersionUID

commit 5f0f31eab63b0e6f6105708ded7f86c96c830781
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Thu Dec 7 13:35:33 2017 -0500

    annotated intervals kebab case
    updated germline WDL workflows

commit 29cc6234dbfb8db12559217a650c6ceb170c5797
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Thu Dec 7 13:15:28 2017 -0500

    cleanup test files

commit 08a35bb4e65eceb735adcd41a91132e9a34d2b66
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Thu Dec 7 02:50:19 2017 -0500

    update WDL scripts

commit 12bcfa192ee6fa6da21239ebf5b513633efe974f
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Thu Dec 7 02:47:33 2017 -0500

    significant updates to GermlineCNVCaller
    integration tests for GermlineCNVCaller w/ sim data in both run modes

commit 151416a4af735ca721bd75e4b54a780c17ac9397
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Thu Dec 7 01:42:05 2017 -0500

    hybrid ADVI abstract argument collection w/ flexible default values
    hybrid ADVI argument collection for contig ploidy model
    hybrid ADVI argument collection for germline denoising and calling model

commit 56e21bf955d3dc0c52aceb384f28cf6173959de0
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Wed Dec 6 23:18:39 2017 -0500

    rewritten python-side coverage metadata table reader using pandas to fix the issues with comment line
    change criterion for cohort/case based on whether a contig-ploidy model is provided or not
    simulated test files for ploidy determination tool
    proper integration test for ploidy determination tool and all edge cases
    updated docs for ploidy determination tool

commit 7fa104b2e9170770cfc5b338835e41215d7fd39c
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Wed Dec 6 18:43:17 2017 -0500

    kabab case for gCNV-related tools
    removed short args (this also partially affected PlotDenoisedCopyRatios and PlotModeledSegments and their integration tests)

commit f02cb024331a986213cfd9fae2da706bbc5ddbd9
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Wed Dec 6 18:02:40 2017 -0500

    synced with mb_gcnv_python_kernel

commit 2963bbf8c90418d9b88545c93771ae51cf542db9
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Wed Dec 6 11:38:05 2017 -0500

    Fixing typo in travis.yml

commit 6cf589999c716ec66404eb0a2ae4310dd130a772
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Wed Dec 6 11:13:58 2017 -0500

    editable, full path

commit d998f2d5c2b33dd41e291be9bfeaea72fe479b8a
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Wed Dec 6 10:56:24 2017 -0500

    revert Dockerfile, change yml

commit 930d7486b7d2cf918fcb16dd03394bb9c9f0611b
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Wed Dec 6 10:34:46 2017 -0500

    more Dockerfile

commit 94112131526b514ef254bcc2c50a239dbae35aa1
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Wed Dec 6 10:25:13 2017 -0500

    more Dockerfile

commit 7d2646240a65f5c0f09f5f25f3e19e9d9bf004d9
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Wed Dec 6 10:06:11 2017 -0500

    more Dockerfile

commit f1235c25aeba85570b5ce389a34975f1b7b5ec3c
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Wed Dec 6 09:39:46 2017 -0500

    Dockerfile edit

commit 3df84dd4693f28e4e8b34fd33f877e99749dffce
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 5 16:08:06 2017 -0500

    Update test PoNs

commit 2c3b20e62a1cba7af24c0b0846eb1629422f51e6
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 5 15:49:38 2017 -0500

    Update test files

commit c65c6e9144ef396792364ab2e06b7b436bb97684
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 5 15:30:59 2017 -0500

    Adding no-GC/do-GC WDL tests.

commit 56451843066a456d9cf8e6eac55ae4df2c518ec3
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 5 12:51:17 2017 -0500

    Updates to handle SAM header changes from sl_wgs_acnv_headers and updates to mb_gcnv_python_kernel.

commit d02d04df684a2820308a1d1c2bfda4b7d1c5f05e
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Mon Nov 13 12:52:33 2017 -0500

    Added CLIs and WDL for python gCNV pipeline.

commit 66ed74b68375d43514ef84658e7a6c771ed9053c
Author: Mehrtash Babadi <mehrtash@broadinstitute.org>
Date:   Wed Nov 15 01:50:03 2017 -0500

    Polished code, ready for review
    
    gCNV computational kernel (initial release)
    
    renaming gammas_s to psi_s to uniformity (sample-specific unexplained variance)
    
    renamed determine_ploidy_and_depth.py to cohort_determine_ploidy_and_depth.py
    finite-temperature forward-backward algorithm
    in the ploidy model, replaced alpha_j (NB over-dispersion) with psi_j (unexplained variance) for uniformity. Also, added the possibility of sample-specific unexplained variance in the germline contig ploidy model
    
    updated I/O routines and CLIs according to team discussion
    
    updated I/O routines and CLIs according to team discussion
    
    changed the output layout of the ploidy determination tool
    refactored parts of io.py
    upped the version to 0.3 as it is not backwards compatible anymore
    
    case ploidy determination tool from a given ploidy model
    major code cleanup and refactoring of I/O module
    refactoring of common CLI script snippets
    
    removed all "targets"
    some code cleanup
    
    pad flat class bitmask w/ a given padding value in the hybrid q_c_expectation_mode
    option to disable annealing and keep the temperature fixed
    
    bugfix in finite-temperature forward-backward
    further refactoring of model I/O
    
    the option to take a previously trained model as starting point in cohort CLI
    the option to take previous calls as a starting point in cohort CLI
    
    option to save and load adamax moments
    
    import/export adamax bias correction tensor
    
    refactoring related to fancy opt I/O
    added average ploidy column to read depth
    updated docs of hybrid inference
    
    modeling intervals can span multiple contigs now; ploidy can change
    across contigs with no issue
    
    save/load adamax state to .npy instead of .tsv for speed
    
    part 1 of doc updates
    
    part 2 of doc updates
    
    part 3 of doc updates
    
    part 4 of doc updates
    
    bumped version to 0.5
    readme
    
    update readme
    
    last minute stylistic doc updates.
````

samuelklee commented:
Split into commits to make things easier to review.  The first commit, which is from sl_wgs_acnv_headers_docs, can be ignored and should be removed after that branch is merged.  This should be ready for review, but the tests may fail if they run too long; we can trim them more after review.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidy.java:
```diff
@@ -0,0 +1,285 @@
+ * @author Samuel Lee &lt;slee@broadinstitute.org&gt;
+ */
+@CommandLineProgramProperties(
+        summary = "Model the baseline ploidy per contig for germline samples given their fragment counts.",
```
`fragment counts` -> just `counts` here and below.

mbabadi replied:
Done.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidy.java:
```diff
@@ -0,0 +1,285 @@
+import java.util.stream.Collectors;
+
+/**
+ * Model the baseline ploidy per contig for germline samples given their fragment counts.
```
`fragment counts` -> just `counts` here and below, to be consistent with docs in the ModelSegments pipeline. It's possible that we might add non-fragment-based counts at some point, so no need to be too specific.

mbabadi replied:
Done.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:
```diff
@@ -0,0 +1,408 @@
+                "the output of DetermineGermlineContigPloidy.",
+        programGroup = CopyNumberProgramGroup.class
+)
+@DocumentedFeature
```
Add `@BetaFeature`.

mbabadi replied:
Done.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:
```diff
@@ -0,0 +1,408 @@
+            Utils.validateArg(inputReadCountFiles.size() > 1, "At least two samples must be provided in the " +
+                    "COHORT mode");
+            if (inputModelDir != null) {
+                logger.info("(advanced feature) A denoising-model directory is provided in the COHORT run mode; " +
```
`run mode` -> `mode` for consistency.

mbabadi replied:
Done.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:
```diff
@@ -0,0 +1,408 @@
+        } else { // case run-mode
+            logger.info("Running the tool in the CASE mode...");
+            Utils.validateArg(inputModelDir != null, "An input denoising-model directory must be provided in the " +
+                    "CASE run-mode.");
```
`run-mode` -> `mode` here and elsewhere.

mbabadi replied:
Done.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:
```diff
@@ -0,0 +1,408 @@
+    }
+
+    private List<File> writeIntervalSubsetReadCountFiles() {
+        logger.info("Validating and aggregating metadata from input read-count files...");
```
`metadata` -> `data`

mbabadi replied:
Done.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:
```diff
@@ -0,0 +1,408 @@
+ *   -L intervals.interval_list \
+ *   --input normal_1.counts.hdf5 \
+ *   --input normal_2.counts.hdf5 \
+ *   ... \
```
You are missing `--contig-ploidy-calls`.

mbabadi replied:
Done.

mbabadi replied:
@samuelklee does `-L` have a long name? I couldn't find it and using short argument names clashes with my sense of aesthetics haha :)

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:
```diff
@@ -0,0 +1,408 @@
+ *   -L intervals.interval_list \
+ *   --model previous_model_path \
+ *   --input normal_1.counts.hdf5 \
+ *   ... \
```
You are missing `--contig-ploidy-calls`.

sooheelee replied:
And `--run-mode`.

mbabadi replied:
Done, Done.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:
```diff
@@ -0,0 +1,408 @@
+import java.util.stream.Collectors;
+
+/**
+ * Call copy-number variants in germline samples given their fragment counts and the corresponding output of
```
`fragment counts` -> `counts`

mbabadi replied:
Done.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:
```diff
@@ -0,0 +1,408 @@
+ * @author Samuel Lee &lt;slee@broadinstitute.org&gt;
+ */
+@CommandLineProgramProperties(
+        summary = "Denoise and call copy-number variants in germline samples given their fragment counts and " +
```
`fragment counts` -> `counts` here and below.

mbabadi replied:
Done.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:
```diff
@@ -0,0 +1,408 @@
+    @Argument(
+            doc = "Input annotated-interval file containing annotations for GC content in genomic intervals " +
+                    "(output of AnnotateIntervals).  All intervals specified via -L must be contained.  " +
+                    "This input should not be provided if an input denoising-model directory is given (that latter " +
```
`that` -> `the`?

mbabadi replied:
Done.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/arguments/GermlineCallingArgumentCollection.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Argument(
+            doc = "Prior probability of alt copy number with respect to contig baseline state in the reference copy number.",
+            fullName = "p-alt",
```
Perhaps extract these to `public static final` in this class.

mbabadi replied:
Done.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/arguments/GermlineContigPloidyModelArgumentCollection.java:
```diff
@@ -0,0 +1,74 @@
+    @Argument(
+            doc = "Contig-level mean bias standard deviation.  If a single sample is provided, " +
+                    "this input will be ignored.",
+            fullName = "mean-bias-standard-deviation",
```
Perhaps extract these to `public static final` in this class.

mbabadi replied:
Done.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/arguments/GermlineDenoisingModelArgumentCollection.java:
```diff
@@ -0,0 +1,159 @@
+
+    @Argument(
+            doc = "Maximum number of bias factors.",
+            fullName = "max-bias-factors",
```
Extract these to `public static final` in this class.

mbabadi replied:
Done.

samuelklee commented on src/test/java/org/broadinstitute/hellbender/GATKBaseTest.java:
```diff
@@ -29,6 +29,7 @@
 
     public static final String packageRootTestDir = publicTestDir + "org/broadinstitute/hellbender/";
     public static final String toolsTestDir = packageRootTestDir + "tools/";
+    public static final String copyNumberTestDir = toolsTestDir + "copynumber/";
```
Let's not add this.

mbabadi replied:
OK.

samuelklee commented on src/test/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidyIntegrationTest.java:
```diff
@@ -0,0 +1,102 @@
+public final class DetermineGermlineContigPloidyIntegrationTest extends CommandLineProgramTest {
+    private static final String gCNVSimDataDir = copyNumberTestDir + "gcnv-sim-data/";
+    private static final File testContigPloidyPriorFile =
+            new File(gCNVSimDataDir + "contig_ploidy_prior.tsv");
```
`new File(gCNVSimDataDir, "contig_ploidy_prior.tsv")`

mbabadi replied:
Done.

samuelklee commented on src/test/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidyIntegrationTest.java:
```diff
@@ -0,0 +1,102 @@
+    private static final File testContigPloidyPriorFile =
+            new File(gCNVSimDataDir + "contig_ploidy_prior.tsv");
+    private static final File[] testCountFiles = IntStream.range(0, 20)
+            .mapToObj(n -> new File(gCNVSimDataDir + String.format("SAMPLE_%03d_counts.tsv", n)))
```
`new File(gCNVSimDataDir, String.format("SAMPLE_%03d_counts.tsv", n))`

samuelklee commented on src/test/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCallerIntegrationTest.java:
```diff
@@ -0,0 +1,83 @@
+public final class GermlineCNVCallerIntegrationTest extends CommandLineProgramTest {
+    private static final String gCNVSimDataDir = copyNumberTestDir + "gcnv-sim-data/";
+    private static final File[] testCountFiles = IntStream.range(0, 20)
+            .mapToObj(n -> new File(gCNVSimDataDir + String.format("SAMPLE_%03d_counts.tsv", n)))
```
Same thing about `new File...` here and below.

mbabadi replied:
Done.

samuelklee commented:
Note that gCNV WDL tests are now passing, but at ~1:05 come in at just under the Travis time limit of ~1:10.  They also run twice as long as the other tests, so it would be good to trim even further.

samuelklee commented:
@sooheelee There is a bit of Javadoc for you to review in the "Added Java CLIs for python gCNV" commit.

sooheelee commented:
Ok @samuelklee. I will take a look.

samuelklee commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/bin/case_denoising_calling.py:
```diff
@@ -0,0 +1,210 @@
+gcnvkernel.HybridInferenceParameters.expose_args(parser)
+
+
+def update_args_dict_from_exported_model(input_model_path: str,
```
A general comment: Use of type hinting is great but somewhat inconsistent.  Can you file an issue to go back and clean up?

mbabadi replied:
Will do :) it is hard to be too consistent since using most external modules (even `numpy`) breaks type inference. One conservative (though, verbose) strategy is to hint _everything_, as if we are coding in a statically typed language.

cmnbroad commented on scripts/gatkcondaenv.yml:
```diff
@@ -17,21 +22,28 @@ dependencies:
+  - tqdm==4.19.4
   - werkzeug==0.12.2
-
+  - "--editable=/gatk/src/main/python/org/broadinstitute/hellbender/tools/gcnv/"
```
@samuelklee The env changes mostly seem fine (@lucidtronix - are you using agparse ?), except for the last line. Rather than the conda .yml be dependent on the repo structure, we should remove that last line, and instead rely on the changes implemented in https://github.com/broadinstitute/gatk/pull/3964. It basically bundles up everything under src/main/python/org/broadinstitute/hellbender/ into a zip file which is then pip installed as part of the conda env, along with all of the other dependencies.

The only reason that https://github.com/broadinstitute/gatk/pull/3964 currently says "DO NOT MERGE" is because I included a commit with a dummy python package, and a corresponding test, so I could validate that it works on travis. Let me know if you want to cherry-pick the core change, or we could merge that PR as is if you want to rebase on top of it.

So, I think most python code should live under src. You could also have tool-specific code that is kept as a tool resource to make it easy to call from Java, but that would be optional.

samuelklee replied:
OK, that all makes sense.  I’m fine with merging that PR and rebasing this one on top of it.

cmnbroad replied:
Ok, its merged.

samuelklee replied:
I pushed a copy of this branch that is rebased on master to https://github.com/broadinstitute/gatk/tree/sl_gcnv_ploidy_cli_rebased to make sure we can use the changes by @cmnbroad mentioned above.  @mbabadi see the last commit in that branch for an example of how we might reorganize things.

I removed the duplicate copies of the gcnv/bin scripts (since these are already in src/main/resources), moved the gcnvkernel package to the same level as the gatkpython dummy package, and used find_packages() to install both as a single gatkpythonpackages package.  I retained a (slightly modified) copy of @mbabadi's original setup.py for gcnvkernel, which one can use to install that package separately should one choose to.

Not sure if this is how we were planning to organize things (and I must admit that I'm not familiar with the subtleties and conventions of python packaging), but it seems to work---we'll see if tests pass in that copy of the branch.

samuelklee replied:
Looks like tests run successfully in that branch, so what I did indeed works.  (Unfortunately, the tests don't pass, but that's because they are still running long...)

mbabadi replied:
I'll come back to this in a bit.

mbabadi replied:
quick note @cmnbroad: `argparse` is a core python3 package now and does not need to be installed separately.

cmnbroad replied:
Oh ok, got it.

sooheelee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidy.java:
```diff
@@ -0,0 +1,285 @@
+import htsjdk.samtools.SAMSequenceDictionary;
+import org.broadinstitute.barclay.argparser.Advanced;
+import org.broadinstitute.barclay.argparser.Argument;
+import org.broadinstitute.barclay.argparser.ArgumentCollection;
```
This appears to be a new tool. Does it need the BetaFeature tag?
```
import org.broadinstitute.barclay.argparser.BetaFeature;
```

mbabadi replied:
Added.

sooheelee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidy.java:
```diff
@@ -0,0 +1,285 @@
+ *      path and must be not provided again.</dd>
+ * </dl>
+ *
+ * <h3>Examples</h3>
```
Examples --> Usage examples

mbabadi replied:
Done.

sooheelee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidy.java:
```diff
@@ -0,0 +1,285 @@
+ *
+ * <p>CASE run-mode:</p>
+ * <pre>
+ * gatk-launch --javaOptions "-Xmx4g" DetermineGermlineContigPloidy \
```
See my comment above on Xmx

sooheelee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:
```diff
@@ -0,0 +1,408 @@
+ * than 10000 consecutive intervals spanning at least 10 - 50 mb.</p></dd>
+ * </dl>
+ *
+ * <h3>Examples</h3>
```
--> Usage examples

mbabadi replied:
Done.

sooheelee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:
```diff
@@ -0,0 +1,408 @@
+ *
+ * <p>COHORT run-mode:</p>
+ * <pre>
+ * gatk-launch --javaOptions "-Xmx4g" GermlineCNVCaller \
```
See my comment on Xmx above. It would be great to have some rule of thumb on how much memory is needed for X number of WES/WGS samples.
  
Tool doc says `--run-mode` is also a required argument so we should include these in the example commands.

mbabadi replied:
@sooheelee I will give an example and mention that memory requirements scale linearly with the number of samples and targets. Let's cover this properly in the tutorial, or a forum post, properly.

mbabadi replied:
@sooheelee Also, the memory allocation of this tool is done outside of JVM and by the spawned python process. The user must ensure that the _machine_ has enough memory.

sooheelee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:
```diff
@@ -0,0 +1,408 @@
+ * </dl>
+ *
+ * <h3>Important Remarks</h3>
+ * <dl>
```
It would be great if the usage example commands showcased parameters we expect users to commonly tweak, given certain types of data. Can we that the default parameters are meant for whole-genome sequences? What would a WES command look like?

samuelklee replied:
We're still running internal evaluations to determine the best values of parameters, many of which determine priors for the gCNV model.  We'll likely set tweak some of the default values to work well for data generated at the Broad, but as @mbabadi warns elsewhere in the documentation, there is no getting around the user having to do some experimentation to set these for their own data.

More generally, as our tools and models become more complex, we cannot expect that they'll run out of the box with one-size-fits-all parameters on all data.  Hopefully we can avoid giving this impression in documentation as well.

sooheelee replied:
Empirical example commands of what we use in production, plus a general description of the data type, would be great to showcase in the usage examples.

mbabadi replied:
@sooheelee @samuelklee While I agree with the complexity of the models and the inevitable non-universality of parameters, I am still hopeful that we can find _decent_ defaults for WGS, WGS, and gene panel data. It requires extensive evaluation on Broad and non-Broad data and is beyond the scope of the first release IMO. At least, we will try to ship the first version with non-garbage default parameters :D

sooheelee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:
```diff
@@ -0,0 +1,408 @@
+import org.broadinstitute.barclay.argparser.Advanced;
+import org.broadinstitute.barclay.argparser.Argument;
+import org.broadinstitute.barclay.argparser.ArgumentCollection;
+import org.broadinstitute.barclay.argparser.CommandLineProgramProperties;
```
Is this tool out of Beta status now?

samuelklee replied:
This is a complete python-based rewrite of the old Spark-based GermlineCNVCaller, so it should still be marked Beta.

mbabadi replied:
Done.

sooheelee commented:
I've put in my two cents for the two tool docs @samuelklee.

samuelklee commented:
Great, thanks!  Also, I think I comments don’t that both tools will indeed be beta; @mbabadi will add these tags.

samuelklee commented:
Great, thanks!  Also, I think I commented above that both tools will indeed be beta; @mbabadi will add these tags.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidy.java:
```diff
@@ -0,0 +1,285 @@
+ *
+ * <p>COHORT run-mode:</p>
+ * <pre>
+ * gatk-launch --javaOptions "-Xmx4g" DetermineGermlineContigPloidy \
```
This will probably scale linearly with both the number of samples and intervals, but may also depend on the complexity of the model (e.g., maximum allowed copy number).  We'll try to come up with some recommendations, but let's leave it for after release.

mbabadi replied:
I believe 4g is more than enough for the ploidy tool for all use cases within reason (up to thousands of samples at the same time).

samuelklee commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/tasks/inference_task_base.py:
```diff
@@ -0,0 +1,757 @@
+import numpy as np
```
Filed #4043.

samuelklee commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/tasks/inference_task_base.py:
```diff
@@ -0,0 +1,757 @@
+        _logger.debug('The {0} for epoch {1} successfully finished in {2:.2f}s'.format(
+            task_name, i_epoch, self._t1 - self._t0))
+
+    def _log_interrupt(self, task_name: str, i_epoch: int):
```
Just curious, why do you take special precautions for `KeyboardInterrupt`?  Is this a TQDM thing?

mbabadi replied:
It is very useful for interactive mode. I don't think it would have any effect if the scripts are run via `PythonScriptExecutor` since `stdin` is not piped (is it @cmnbroad?)

cmnbroad replied:
Right. Technically I think you could write to stdin and the underlying process controller would honor it, but the PythonScriptExecutor doesn't rely on it (since you always run a script, module, or command line command). The StreamingPythonScriptExecutor does so though.

mbabadi replied:
@cmnbroad Perfect, thank you!

samuelklee commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/models/theano_hmm.py:
```diff
@@ -0,0 +1,211 @@
+import numpy as np
```
See #4043.  Let's also add some tests for the HMM.

mbabadi replied:
Edited your issue.

samuelklee commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/models/model_ploidy.py:
```diff
@@ -0,0 +1,364 @@
+        return validated_contig_ploidy_prior_map, num_ploidy_states
+
+    @staticmethod
+    def get_contig_ploidy_prior_map_from_tsv_file(input_path: str,
```
This seems like it should be in io_ploidy.py.

mbabadi replied:
Done.

samuelklee commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/models/model_ploidy.py:
```diff
@@ -0,0 +1,364 @@
+        return PloidyModelConfig(**relevant_kwargs)
+
+
+class PloidyWorkspace:
```
Just a thought: if we split off the inference package, it would be useful to have a simple model as a pedagogical example for developers.  Such a model could highlight tricks like the sharing of tensors here.

mbabadi replied:
I completely agree, and I think it is a good idea to do so. Will make an issue.

mbabadi replied:
Oh nevermind, you already made it.

samuelklee commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/models/model_denoising_calling.py:
```diff
@@ -0,0 +1,1168 @@
+        return fact * np.ones((self.denoising_model_config.max_bias_factors,), dtype=types.floatX)
+
+
+class DenoisingModel(GeneralizedContinuousModel):
```
Just wanted to say, this looks great!

mbabadi replied:
Thank you!! ;-)

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidy.java:
```diff
@@ -0,0 +1,285 @@
+ * Model the baseline ploidy per contig for germline samples given their fragment counts.
+ * These should be either HDF5 or TSV count files generated by {@link CollectFragmentCounts}.
+ *
+ * <p>This tool has two run-modes as described below:</p>
```
Fixed.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidy.java:
```diff
@@ -0,0 +1,285 @@
+ *     <dt>COHORT run-mode</dt>
+ *     <dd>If a model parameter directory is not provided, the tool is run in the COHORT mode. In this run-mode,
+ *      ploidy model parameters (e.g. coverage bias and variance in each contig) are inferred
+ *      along with baseline contig ploidy of each sample. A table specifying priors for the ploidy per contig
```
Haha, done :)

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidy.java:
```diff
@@ -0,0 +1,285 @@
+ * <p>This tool has two run-modes as described below:</p>
+ * <dl>
+ *     <dt>COHORT run-mode</dt>
+ *     <dd>If a model parameter directory is not provided, the tool is run in the COHORT mode. In this run-mode,
```
Done. Also, gave an example about the format of the prior table.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidy.java:
```diff
@@ -0,0 +1,285 @@
+ *      coverage variance estimates.</dd>
+ *
+ *     <dt>CASE run-mode</dt>
+ *     <dd>If a previously obtained model parameter bundle is provided (e.g. from a previous run on a similar set
```
Done.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidy.java:
```diff
@@ -0,0 +1,285 @@
+ * gatk-launch --javaOptions "-Xmx4g" DetermineGermlineContigPloidy \
+ *   --input normal_1.counts.hdf5 \
+ *   --input normal_2.counts.hdf5 \
+ *   ... \
```
Good catch!

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidy.java:
```diff
@@ -0,0 +1,285 @@
+        oneLineSummary = "Model the baseline ploidy per contig for germline samples given their fragment counts.",
+        programGroup = CopyNumberProgramGroup.class
+)
+@DocumentedFeature
```
Done.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidy.java:
```diff
@@ -0,0 +1,285 @@
+
+        if (inputModelDir != null) {
+            runMode = RunMode.CASE;
+            logger.info("A contig-ploidy model was provided, running in case runMode...");
```
yeah, that's exactly how it happened. Fixed them all.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/DetermineGermlineContigPloidy.java:
```diff
@@ -0,0 +1,285 @@
+            script = CASE_DETERMINE_PLOIDY_AND_DEPTH_PYTHON_SCRIPT;
+            arguments.add("--input_model_path=" + inputModelDir);
+        }
+        System.out.println(arguments.stream().collect(Collectors.joining(" ")));
```
Done.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:
```diff
@@ -0,0 +1,408 @@
+ *
+ * <p>CASE run-mode:</p>
+ * <pre>
+ * gatk-launch --javaOptions "-Xmx4g" GermlineCNVCaller \
```
Done.

mbabadi commented on scripts/cnv_wdl/germline/README.md:
```diff
@@ -0,0 +1,53 @@
+- ``CNVGermlineCohortWorkflow.contig_ploidy_priors`` -- TSV file containing prior probabilities for the ploidy of each contig, with column headers: CONTIG_NAME, PLOIDY_PRIOR_0, PLOIDY_PRIOR_1, ...
+- ``CNVGermlineCohortWorkflow.gatk_docker`` -- GATK Docker image (e.g., ``broadinstitute/gatk:latest``).
+- ``CNVGermlineCohortWorkflow.intervals`` -- Picard or GATK-style interval list.  For WGS, this should typically only include the chromosomes of interest.
+- ``CNVGermlineCohortWorkflow.normal_bams_list`` -- TSV file consisting of corresponding BAM and index paths as described in cnv_germline_cohort_workflow.wdl.
```
I'll leave it for you then.

mbabadi commented on scripts/cnv_wdl/germline/cnv_germline_case_workflow.wdl:
```diff
@@ -0,0 +1,355 @@
+#   For WGS, the intervals should simply cover the chromosomes of interest.
+#
+# - Example invocation:
+#    java -jar cromwell.jar run cnv_germline_case_workflow.wdl myParameters.json
```
Done.

mbabadi commented on scripts/cnv_wdl/germline/cnv_germline_case_workflow.wdl:
```diff
@@ -0,0 +1,355 @@
+    File ref_fasta_fai
+    File ref_fasta
+    String gatk_docker
+    Int? preemptible_attempts
```
I don't think so (it defaults to 2). Setting it to 0 asks for a non-preemptible machine.

samuelklee replied:
Sorry, I meant that this should be moved to the section containing optional arguments.

mbabadi replied:
Oh, OK! :)

mbabadi commented on scripts/cnv_wdl/germline/cnv_germline_case_workflow.wdl:
```diff
@@ -0,0 +1,355 @@
+                disable_annealing = gcnv_disable_annealing,
+                preemptible_attempts = preemptible_attempts
+        }
+    }
```
@samuelklee @asmirnov239 We can exclude intermediate .tar.gz outputs iff we include _everything_ properly in the (g)VCF output. Perhaps we can make this a mandate, assume that it will be so in the next maintenance release (if not for the release), and remove intermediate .tar.gz for now.

mbabadi replied:
I'll make an issue and leave this for now.

mbabadi commented on scripts/cnv_wdl/germline/cnv_germline_cohort_workflow.wdl:
```diff
@@ -0,0 +1,428 @@
+    File ref_fasta_fai
+    File ref_fasta
+    String gatk_docker
+    Int? preemptible_attempts
```
See above.

mbabadi commented on scripts/cnv_wdl/germline/cnv_germline_cohort_workflow.wdl:
```diff
@@ -0,0 +1,428 @@
+                disable_annealing = gcnv_disable_annealing,
+                preemptible_attempts = preemptible_attempts
+        }
+    }
```
In another issue.

mbabadi commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/io/io_commons.py:
```diff
@@ -0,0 +1,511 @@
+import logging
```
I agree. Let's do it all with pandas.

mbabadi commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/preprocess/interval_list_mask.py:
```diff
@@ -0,0 +1,98 @@
+                self.drop_t[ti] = True
+                self.drop_reason_t[ti].add("cohort-wide uncovered interval")
+
+    def drop_intervals_with_anomalous_coverage(self):
```
Removed.

mbabadi commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/structs/interval.py:
```diff
@@ -0,0 +1,149 @@
+        return padded_interval
+
+    def overlaps_with(self, other):
+        assert isinstance(other, Interval), "{0} is not of Interval type".format(other)
```
Removed. That's a main caveat of dynamic typing. No checks and you may get weird errors; and no one forces you to check so the get codes inconsistent...

mbabadi commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/structs/interval.py:
```diff
@@ -0,0 +1,149 @@
+        else:
+            return np.abs(self.get_midpoint() - other.get_midpoint())
+
+    def __eq__(self, other):
```
I do not assume well-ordered intervals in `gcnvkernel` (no sort op, etc.). In the evaluation suite, I rely on `IntervalTree`'s own ordering. So I think I can safely drop the comparison ops.

mbabadi commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/structs/interval.py:
```diff
@@ -0,0 +1,149 @@
+        return "GC_CONTENT"
+
+
+class NameAnnotation(IntervalAnnotation):
```
No harm, but I will drop it to make you happy :D

mbabadi commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/structs/metadata.py:
```diff
@@ -0,0 +1,294 @@
+_logger = logging.getLogger(__name__)
+
+
+class IntervalListMetadata:
```
Will make an issue.

mbabadi commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/structs/metadata.py:
```diff
@@ -0,0 +1,294 @@
+                              "coverage metadata can result in unreliable ploidy designations. It is recommended " \
+                              "that the user verifies this designation by orthogonal methods."
+
+        def rectify_contig(_contig: str):
```
I agree. Though, I don't think any reference uses anything other than `X` and `chrX` (past and future). Ideally, `check_germline_contig_ploidy_for_homo_sapiens` must be refactored to take a data structure representing sex genotypes and associated germline contig ploidy states + table that could be optionally passed to python (and Java) to generate these warnings. I'd say too much work (and no practical gain) for now.

mbabadi commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/tasks/task_case_denoising_calling.py:
```diff
@@ -0,0 +1,148 @@
+        return self.copy_number_update_reduced
+
+
+class CopyNumberEmissionSampler(Sampler):
```
Yep. Removed.

mbabadi commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/utils/cli_commons.py:
```diff
@@ -0,0 +1,63 @@
+    """Configures python logger according to parsed arguments."""
+    logging.basicConfig(level=log_level_map[parsed_args.logfile_log_level],
+                        format='%(asctime)s %(name)-12s %(levelname)-8s %(message)s',
+                        datefmt='%m-%d %H:%M',
```
Added.

mbabadi commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/setup.py:
```diff
@@ -0,0 +1,38 @@
+
+assert sys.version_info >= (3, 4), "gcnvkernel requires Python 3.4.x or later"
+
+VERSIONFILE="gcnvkernel/_version.py"
```
True, I just took a template and edited. Fixed PEP8 violations.

mbabadi commented on src/main/resources/org/broadinstitute/hellbender/tools/copynumber/case_denoising_calling.py:
```diff
@@ -0,0 +1,210 @@
+import os
```
The easiest solution is to let the scripts be treated as Java resources and not distributing them with gcnvkernel. I will come back to this in a bit.

mbabadi commented on src/main/resources/org/broadinstitute/hellbender/tools/copynumber/cohort_determine_ploidy_and_depth.py:
```diff
@@ -0,0 +1,119 @@
+                   type=str,
+                   required=True,
+                   default=argparse.SUPPRESS,
+                   help="Output path to write the ploidy model for future single-sample ploidy determination use")
```
Done.

mbabadi commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/models/commons.py:
```diff
@@ -0,0 +1,244 @@
+                              mu > 0, value >= 0, alpha > 0)
+
+
+def negative_binomial_gaussian_approx_logp(mu, alpha, value):
```
Huh, this was pretty useful, thanks! the confidence levels are interesting -- it is impossible to know for sure whether an apparently unused attribute is truly unused. In fact, it has made a quite a few errors.

mbabadi replied:
I removed `poisson_logp` but kept `negative_binomial_gaussian_approx_logp` (will make an issue about it -- there's a potential for 2~3x speedup if we can pull it off with the right hack).

mbabadi commented on src/main/python/org/broadinstitute/hellbender/tools/gcnv/gcnvkernel/models/dists.py:
```diff
@@ -0,0 +1,63 @@
+import theano.tensor as tt
+
+
+class PositiveFlatTop(PositiveContinuous):
```
Not anymore, but it is a useful distribution and I think I will end up using it at some point. I'll keep it here for now.

mbabadi commented:
@samuelklee Thanks for the PR review, I have implemented most changes and will make issues for the remaining momentarily. Waiting for the travis tests to pass.

mbabadi commented:
@sooheelee Thank you for the docs review! I applied all of the changes. Will coordinate with you regarding the tutorial and memory allocation tips after the release.

samuelklee commented:
OK, just a few more minor doc comments from me.  Also the issue about preemptible_attempts above.  Otherwise good to rebase and merge!  I will take care of trimming tests and reorganizing the packaging in another PR as we discussed.

samuelklee commented:
@asmirnov239 @MartonKN Comment now (or hold your peace until after release!)  We will merge soon otherwise.

droazen commented:
@samuelklee Note that `--javaOptions` is now `--java-options`

samuelklee commented:
Thanks for catching that @droazen, but I think we removed those for now.
